lib/rules/no-constant-binary-expression.js

```diff
@@ -0,0 +1,437 @@
+/**
+ * @fileoverview Rule to flag constant comparisons and logical expressions that always/never short circuit
+ * @author Jordan Eldredge <https://jordaneldredge.com>
+ */
+
+"use strict";
+
+const { isNullOrUndefined } = require("./utils/ast-utils");
+
+// Object keys which, if redefined, could impact how equality is computed.
+const OBJECT_EQUALITY_KEYS = new Set(["valueOf", "toString", "__proto__"]);
+const NUMERIC_OR_STRING_BINARY_OPERATORS = new Set(["+", "-", "*", "/", "%", "|", "^", "&"]);
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+/**
+ * Test if an AST node has a statically knowable constant truthiness. Meaning,
+ * it will always coerce to either `true` or `false`.
+ * @param {ASTNode} node The AST node being tested.
+ * @returns {boolean} Does `node` have constant truthiness?
+ */
+function hasConstantTruthiness(node) {
+    switch (node.type) {
+        case "ObjectExpression": // Objects are always truthy
+        case "ArrayExpression": // Arrays are always truthy
+        case "ArrowFunctionExpression": // Functions are always truthy
+        case "FunctionExpression": // Functions are always truthy
+        case "ClassExpression": // Classes are always truthy
+        case "NewExpression": // Objects are always truthy
+        case "Literal": // Truthy, or falsy, literals never change
+            return true;
+        case "CallExpression": {
+            if (node.callee.type === "Identifier" && node.callee.name === "Boolean") {
+                return node.arguments.length === 0 || hasConstantTruthiness(node.arguments[0]);
+            }
+            return false;
+        }
+        case "JSXElement": // ESLint has a policy of not assuming any specific JSX behavior.
+        case "JSXFragment":
+            return false;
+        case "AssignmentExpression":
+            if (node.operator !== "=") {
+                return false; // We won't go so far as to try to evaluate += etc.
+            }
+            return hasConstantTruthiness(node.right);
+        case "TemplateLiteral":
+
+            /*
+             * TODO: If all quasis are empty, we could look at node.expressions
```

Comment:
Are these TODOs to be completed before merging?

Reply:
These TODOs represent ways in which the rule could technically be taken further, but felt like they passed the point of diminishing returns in my estimation. I wasn't planning to complete these as part of this PR, but if reviewers think any of them seem worth doing, I'd be happy to add them.

My intention was that they would clarify places where I recognized that further inference was technically possible, but had opted not to do that work while also leaving a hint in case someone in the future wanted to consider add it.

Reply:
I'm happy to reword these to be more stylistically consistent with the rest of the project, just let me know what would work.

Reply:
Yeah, I’d maybe say something like “Possible future direction if needed:” just to make it clear that there’s no tech debt here, just an enhancement opportunity.

Reply:
Done.